Invalid bootstrap servers list blocks during startup

```python
>>> kafka.KafkaProducer(bootstrap_servers='can-not-resolve:9092')
# never returns
```

I see your point. So, this is a good thing when there is a temporary problem and the name will eventually be resolved properly again. But what if the provided name is plain wrong? (say, for example, you edited some configuration files and mispelled the server fqdn).

I though "there should be an optional `timeout` argument.". I could not find it at first glance in the documentations (I should say, I am using 1.4.2).  Then, grepping the code for the warning `DNS lookup failed for not-a-valid-name:9092, exception was [Errno -2] Name or service not known. Is your advertised.listeners (called advertised.host.name before Kafka 9) correct and resolvable?`) I found it is produced around

https://github.com/dpkp/kafka-python/blob/c0fddbd24269d4333e3b6630a23e86ffe33dfcb6/kafka/conn.py#L288

(Note that the presence of the named argument made me think I was on the right track)

which is called from 

https://github.com/dpkp/kafka-python/blob/c0fddbd24269d4333e3b6630a23e86ffe33dfcb6/kafka/client_async.py#L245

definitely without any arguments.

In my opinion, this should not enter the infinite loop unavoidably.

I am only starting with both kafka and your code, so I don't know the implications of this (I mean, may be this function is called on bootstrapping and then every 10 seconds for metadata refreshing, I really don't know if what I am asking is way stupid).

This is what I would expect:
```
$ docker run --rm kafka:2.11-1.1.0 kafka-console-producer.sh --topic test --broker-list con-cristina-estabamos-mejor:9092  [19:50:09]
[2018-04-12 22:50:30,316] WARN Removing server con-cristina-estabamos-mejor:9092 from bootstrap.servers as DNS resolution failed for con-cristina-estabamos-mejor (org.apache.kafka.clients.ClientUtils)
org.apache.kafka.common.KafkaException: Failed to construct kafka producer
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:456)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:303)
        at kafka.producer.NewShinyProducer.<init>(BaseProducer.scala:40)
        at kafka.tools.ConsoleProducer$.main(ConsoleProducer.scala:49)
        at kafka.tools.ConsoleProducer.main(ConsoleProducer.scala)
Caused by: org.apache.kafka.common.config.ConfigException: No resolvable bootstrap urls given in bootstrap.servers
        at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:66)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:405)
        ... 4 more
```

--- a/kafka/conn.py
+++ b/kafka/conn.py
@@ -292,11 +292,7 @@ class BrokerConnection(object):
         # First attempt to perform dns lookup
         # note that the underlying interface, socket.getaddrinfo,
         # has no explicit timeout so we may exceed the user-specified timeout
-        while time.time() < timeout:
-            if self._dns_lookup():
-                break
-        else:
-            return False
+        self._dns_lookup()
 
         # Loop once over all returned dns entries
         selector = None
@@ -903,6 +899,7 @@ class BrokerConnection(object):
 
         Returns: version tuple, i.e. (0, 10), (0, 9), (0, 8, 2), ...
         """
+        timeout_at = time.time() + timeout
         log.info('Probing node %s broker version', self.node_id)
         # Monkeypatch some connection configurations to avoid timeouts
         override_config = {
@@ -932,7 +929,7 @@ class BrokerConnection(object):
         ]
 
         for version, request in test_cases:
-            if not self.connect_blocking(timeout):
+            if not self.connect_blocking(timeout_at - time.time()):
                 raise Errors.NodeNotReadyError()
             f = self.send(request)
             # HACK: sleeping to wait for socket to send bytes

